numba/cuda/cudadrv/driver.py

```diff
@@ -2725,18 +2815,21 @@ def add_cu(self, path, name):
                                   "with the ctypes binding. ")
 
     def complete(self):
-        cubin = c_void_p(0)
+        cubin_buf = c_void_p(0)
         size = c_size_t(0)
 
         try:
-            driver.cuLinkComplete(self.handle, byref(cubin), byref(size))
+            driver.cuLinkComplete(self.handle, byref(cubin_buf), byref(size))
         except CudaAPIError as e:
             raise LinkerError("%s\n%s" % (e, self.error_log))
 
         size = size.value
         assert size > 0, 'linker returned a zero sized cubin'
         del self._keep_alive[:]
-        return cubin, size
+
+        # We return a copy of the cubin because it's owned by the linker
+        cubin_ptr = ctypes.cast(cubin_buf, ctypes.POINTER(ctypes.c_char))
+        return bytes(np.ctypeslib.as_array(cubin_ptr, shape=(size,)))
```

Comment:
Does NumPy need to be involved? Could this just be something like `bytes((ctypes.c_char * size)(cubin_ptr))`?

Reply:
I spent a fair amount of time fiddling around to try and work out how to do this without NumPy in the past, and didn't succeed (the suggestion doesn't work, giving `TypeError: incompatible types, LP_c_char instance instead of c_char instance`). Since NumPy is already used in this file, the code is only moved from `codegen.py` (not new in this PR), and it provides a tool to do what's needed, it feels reasonable to me to keep this as-is rather than reimplement something for Numba. Can you see any issue with keeping this as-is?

Reply:
Given it's just a refactor, seems ok to keep as-is, would be good at some point to work out how to resolve this sort of thing though, but not for this PR as working it out it might be quite involved!